src/taxonomy_router.rs

```diff
@@ -0,0 +1,246 @@
+                        let mut payload = String::new();
+                        itry!(req.body.read_to_string(&mut payload));
+                        let selectors: Vec<$sel> =
+                            itry!(serde_json::from_str(&payload));
+
+                        api.$call(selectors)
+                    };
+
+                    self.build_response(&result)
+                }
+            })
+        }
+
+        // Generates the code to process a given HTTP call with a json body.
+        macro_rules! payload_api {
+            ($call:ident, $param:ty, $path:expr, $method:expr) => (
+                if path == $path && req.method == $method {
+                    return {
+                        let api = self.api.lock().unwrap();
+                        let mut payload = String::new();
+                        itry!(req.body.read_to_string(&mut payload));
+                        let selectors: $param =
+                            itry!(serde_json::from_str(&payload));
+
+                        let result = api.$call(selectors);
+                        return self.build_response(&result);
+                    }
+                }
+            )
+        }
+
+        // Generates the code to process a given HTTP call with a json body.
+        // This version takes 2 parameters for the internal call.
+        macro_rules! payload_api2 {
+            ($call:ident, $name1:ident => $param1:ty, $name2:ident => $param2:ty, $path:expr, $method:expr) => (
+                if path == $path && req.method == $method {
+                    return {
+                        let api = self.api.lock().unwrap();
+                        let mut payload = String::new();
+                        itry!(req.body.read_to_string(&mut payload));
+                        #[derive(Deserialize)]
+                        struct Args {
+                            $name1: $param1,
+                            $name2: $param2
+                        }
+                        let args: Args =
+                            itry!(serde_json::from_str(&payload));
+
+                        let result = api.$call(args.$name1, args.$name2);
+                        return self.build_response(&result);
+                    }
+                }
+            )
+        }
+
+        // Keep these urls in sync with the AuthEndpoint(s) in the create() method.
+
+        // Selectors queries.
+        get_post_api!(get_services, ServiceSelector, ["services"]);
+        get_post_api!(get_getter_channels, GetterSelector, ["channels", "getters"]);
+        get_post_api!(get_setter_channels, SetterSelector, ["channels", "setters"]);
+
+        // Fetching and getting values.
+        payload_api!(fetch_values, Vec<GetterSelector>, ["channels", "get"], Method::Post);
+        payload_api!(send_values, TargetMap<SetterSelector, Value>, ["channels", "set"], Method::Post);
+
+        // Adding tags.
```

Comment:
`PUT` usually means "replace" and `POST` "Create/add" so I'll keep them as Post.